### ISSUE TITLE
yum install -y kernel-devel fails with imeout

### ISSUE THREAD

**Issue by ralf-k-a** [dotless-de/vagrant-vbguest#381]
The plugin fails with the plain CentOS/7 box with lots of download timeouts.

In `install_kernel_devel()`, line 72 the CentOS installer triggers installation of the kernel development packages. YUM tries to download from many mirrors, but all downloads fail as too slow. A manual download with `curl` shows, that the downloads start fast and during the download the speed decreases until 0.

In many resources, I found that this is a common problem. Installing epel-release (`yum install -y epel-release`) fixes it. For yet unknown reasons, a short break of 1 second after the installation of epel-release is necessary. Otherwise you will not see any change in behaviour. Maybe human users are not faced with that problem because the short break needed for typing he next command is sufficient.

Attached is my Workaround: [Vagrantfile.txt](https://github.com/dotless-de/vagrant-vbguest/files/5079010/Vagrantfile.txt) contains an extension of the CentOS loader that prepends the fix to the standard installation.

**Comment by fnordfish** [comment]
Hi, thanks for your feedback. I was thinking about this and I wouldn't feel comfortable adding this directly into the project. But maybe we can build something that would make it easier to customize the installer.

We could use the fairly new [`installer_options`](https://github.com/dotless-de/vagrant-vbguest#installer-specific-options-installer_options) config and allow custom scripts to be run prior to the installer.
Something along this:

```ruby
config.vbguest.installer_options = {
  pre_install_script: {
    "CentOS" => "yum install -y epel-release; sleep 1" 
  }
}
```

or

```ruby
config.vm.define "centos" do |box_config|
  # ...
  pre_install_script:  "yum install -y epel-release; sleep 1" 
end
```

There are still some issues with that approach, like mutli box projects with different OSes.

**Comment by kdimiche** [comment]
I just ran into this issue and it's blocking me from using molecule to test against CentOS 8.  Is there an ETA on when it may be fixed?

**Comment by ralf-k-a** [comment]
Hi Robert,


You can add an option, however this option should disable the fix. It is mandatory on CentOS, as far as I understood frpm the many sources in the
Internet I read.

My Fix was just a workaround (I did not have the tim enot te knowledge to provide a real fix).


Greetings


Ralf



Ralf Kretzschmar-Auer
Digital Designer | Senior Solution Architect
CGI Deutschland B.V. & Co. KG
Leinfelder Straße 60 | 70771 Leinfelden-Echterdingen | Germany
T: +49 711 72846 0  | M:+49 178 9277936
Ralf.Kretzschmar-Auer@cgi.com<mailto:Ralf.Kretzschmar-Auer@cgi.com> | de.cgi.com<http://www.de.cgi.com/>
CGI folgen auf: Xing<https://www.xing.com/companies/cgi%22%20%5Co%20%22Google+> | LinkedIn<https://www.linkedin.com/company/cgi%22%20%5Co%20%22LinkedIn> | Twitter<https://twitter.com/cgi_de%22%20%5Co%20%22Twitter> | Facebook<https://www.facebook.com/CGIDeutschland%22%20%5Co%20%22Facebook>
Unsere Pflichtangaben gemäß § 35a GmbHG / §§ 161, 125a HGB finden Sie unter de.cgi.com/pflichtangaben<http://de.cgi.com/pflichtangaben%22%20%5Ct%20%22_blank>.

CONFIDENTIALITY NOTICE: Proprietary/Confidential information belonging to CGI Group Inc. and its affiliates may be contained in this message. If you are not a recipient indicated or intended in this message (or responsible for delivery of this message to such person), or you think for any reason that this message may have been addressed to you in error, you may not use or copy or deliver this message to anyone else. In such case, you should destroy this message and are asked to notify the sender by reply e-mail.

Von: Robert Schulze <notifications@github.com>
Antworten an: dotless-de/vagrant-vbguest <reply@reply.github.com>
Datum: Samstag, 26. September 2020 um 14:54
An: dotless-de/vagrant-vbguest <vagrant-vbguest@noreply.github.com>
Cc: Ralf Kretzschmar-Auer <ralf.kretzschmar-auer@cgi.com>, Author <author@noreply.github.com>
Betreff: Re: [dotless-de/vagrant-vbguest] yum install -y kernel-devel fails with imeout (#381)

EXTERNAL SENDER: Do not click any links or open any attachments unless you trust the sender and know the content is safe.
EXPÉDITEUR EXTERNE: Ne cliquez sur aucun lien et n’ouvrez aucune pièce jointe à moins qu’ils ne proviennent d’un expéditeur fiable, ou que vous ayez l'assurance que le contenu provient d'une source sûre.



Hi, thanks for your feedback. I was thinking about this and I wouldn't feel comfortable adding this directly into the project. But maybe we can build something that would make it easier to customize the installer.

We could use the fairly new installer_options<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dotless-2Dde_vagrant-2Dvbguest-23installer-2Dspecific-2Doptions-2Dinstaller-5Foptions&d=DwMCaQ&c=H50I6Bh8SW87d_bXfZP_8g&r=5FBsl3fYpU5lvIej17Ag8HFRoL9F3KsBqamJ7oFJO00&m=tTwUAhT0jw3Dy7QD1sswK9aLLpCdnh-grSyONU05rOM&s=u42TLqwpHLvaIX3WrjHYoES1UkqGoqVVQKZOGpMb8S4&e=> config and allow custom scripts to be run prior to the installer.
Something along this:

config.vbguest.installer_options = {

  pre_install_script: {

    "CentOS" => "yum install -y epel-release; sleep 1"

  }

}

or

config.vm.define "centos" do |box_config|

  # ...

  pre_install_script:  "yum install -y epel-release; sleep 1"

end

There are still some issues with that approach, like mutli box projects with different OSes.

**Comment by fnordfish** [comment]
Addressed with v0.26.0. See: https://github.com/dotless-de/vagrant-vbguest#installer-hooks-installer_hooks

```ruby
Vagrant.configure("2") do |config|
  config.vm.define "my_cent_os_box" do |c|
    c.vbguest.installer_hooks[:before_install] = ["yum install -y epel-release", "sleep 1"]
  end
end
```
